src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/ProjectNames.cs

```diff
@@ -103,11 +103,73 @@ public static async Task<ProjectNames> FromDTEProjectAsync(EnvDTE.Project dtePro
                 projectId: projectId);
         }
 
-        private static string GetProjectGuid(string fullname, IVsSolution5 vsSolution5)
+        private static string GetProjectGuid(EnvDTE.Project dteProject, SVsSolution vsSolution)
         {
             ThreadHelper.ThrowIfNotOnUIThread();
-            var guid = vsSolution5.GetGuidOfProjectFile(fullname);
-            return guid.ToString();
+
+            // Each project system needs to implement its own DTE project implementation. Hence, custom project types
+            // have the potential to not implement everything. Therefore, we need to try multiple ways to maximise the
+            // chance we find one way that actually works.
+
+            if (TryGetProjectGuidFromUniqueName(dteProject, (IVsSolution2)vsSolution, out Guid guid))
+            {
+                return guid.ToString();
+            }
+
+            if (TryGetProjectGuidFromFullName(dteProject, (IVsSolution5)vsSolution, out guid))
+            {
+                return guid.ToString();
+            }
+
+            throw new ArgumentException("Unable to find project guid");
+
+            bool TryGetProjectGuidFromUniqueName(EnvDTE.Project project, IVsSolution2 vsSolution2, out Guid projectGuid)
+            {
+                try
+                {
+                    var uniqueName = project.UniqueName;
+                    if (string.IsNullOrEmpty(uniqueName))
+                    {
+                        projectGuid = default(Guid);
+                        return false;
+                    }
+
+                    ErrorHandler.ThrowOnFailure(vsSolution2.GetProjectOfUniqueName(uniqueName, out IVsHierarchy hierarchy));
+                    ErrorHandler.ThrowOnFailure(vsSolution2.GetGuidOfProject(hierarchy, out projectGuid));
+
+                    return true;
+                }
+                catch
+                {
+                    projectGuid = default(Guid);
+                    return false;
+                }
+            }
+
+            bool TryGetProjectGuidFromFullName(EnvDTE.Project project, IVsSolution5 vsSolution5, out Guid projectGuid)
+            {
+                try
+                {
+                    var fullName = project.FullName;
+                    if (string.IsNullOrEmpty(fullName))
+                    {
+                        fullName = project.FileName;
+                        if (string.IsNullOrEmpty(fullName))
+                        {
+                            projectGuid = default(Guid);
+                            return false;
+                        }
+                    }
+
+                    projectGuid = vsSolution5.GetGuidOfProjectFile(fullName);
```

Comment:
This probably needs moved to the top method instead right?

Reply:
"unique name" is usually just the relative path to the project file from the solution. Every test I've tried so far returns the exact string that's in the project file. I feel like that's probably more likely to be reliable, hence I preferred it over DTE.FullName. But which of the two is tried first doesn't really matter to me. I can change it if you prefer.

Reply:
All good as long as it's consistent.